File: ads/src/main/java/com/example/ads/newStrategy/GoogleRewarded.java
package com.example.ads.newStrategy;//package com.vyroai.autocutcut.ads.max;
//

import android.content.Context;

import androidx.annotation.NonNull;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.interstitial.InterstitialAd;
import com.google.android.gms.ads.rewarded.RewardedAd;
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Stack;


public class GoogleRewarded {

    private final int totalLevels = 4;
    private ArrayList<ArrayList<Object>> adUnits;
    private final String adUnitId = "ca-app-pub-3940256099942544/1033173712";

    public GoogleRewarded(Context context) {
        instantiateList();
        loadInitialRewards(context);
    }

    private void instantiateList() {
        adUnits = new ArrayList<>();

        adUnits.add(0, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<InterstitialAd>())));
        adUnits.add(1, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<InterstitialAd>())));
        adUnits.add(2, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<InterstitialAd>())));
        adUnits.add(3, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<InterstitialAd>())));
        adUnits.add(4, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<InterstitialAd>())));
    }

    public void loadInitialRewards(Context context) {
        RewardedAdLoad(context, totalLevels);
    }


    public RewardedAd getMediumAd(Context activity) {
        return getRewardedAd(activity, 1);
    }

    public RewardedAd getHighFloorAd(Context activity) {
        return getRewardedAd(activity, 2);
    }

    public RewardedAd getDefaultAd(Context activity) {
        return getRewardedAd(activity, 0);
    }


    public RewardedAd getRewardedAd(Context activity, int maxLevel) {
        for (int i = totalLevels; i >= 0; i--) {

            if (maxLevel > i) {
                break;
            }

            ArrayList<Object> list = adUnits.get(i);
            String adunitid = (String) list.get(0);
            Stack<RewardedAd> stack = (Stack<RewardedAd>) list.get(1);

            RewardedAdLoadSpecific(activity, adunitid, stack);

            if (stack != null && !stack.isEmpty()) {
                return stack.pop();
            }
        }

        return null;
    }

    public void RewardedAdLoad(Context activity, int level) {

        if (level < 0) {
            return;
        }

        if (adUnits.size() < level) {
            return;
        }

        ArrayList<Object> list = adUnits.get(level);
        String adunitid = (String) list.get(0);
        Stack<RewardedAd> stack = (Stack<RewardedAd>) list.get(1);

        AdRequest adRequest = new AdRequest.Builder().build();
        RewardedAd.load(activity, adunitid, adRequest, new RewardedAdLoadCallback() {
            @Override
            public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                super.onAdFailedToLoad(loadAdError);
                RewardedAdLoad(activity, level - 1);
            }

            @Override
            public void onAdLoaded(@NonNull RewardedAd rewardedAd) {
                super.onAdLoaded(rewardedAd);
                stack.push(rewardedAd);
            }
        });
    }

    public void RewardedAdLoadSpecific(Context activity, String adUnitId, Stack<RewardedAd> stack) {
        AdRequest adRequest = new AdRequest.Builder().build();
        RewardedAd.load(activity, adUnitId, adRequest, new RewardedAdLoadCallback() {
            @Override
            public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                super.onAdFailedToLoad(loadAdError);
            }

            @Override
            public void onAdLoaded(@NonNull RewardedAd rewardedAd) {
                super.onAdLoaded(rewardedAd);
                stack.push(rewardedAd);
            }
        });
    }
}

File: ads/src/main/java/com/example/ads/newStrategy/GoogleRewarded.kt
package com.example.ads.newStrategy

import android.content.Context
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback
import java.util.Stack


class GoogleRewarded(context: Context?) {
    private val totalLevels = 4
    private var adUnits: ArrayList<ArrayList<Any>>? = null
    private val reward5 = "ca-app-pub-9507635869843997/1313904170"
    private val reward4 = "ca-app-pub-9507635869843997/8345699462"
    private val rewardHigh = "ca-app-pub-9507635869843997/3561056899"
    private val rewardMed = "ca-app-pub-9507635869843997/1909063878"
    private val rewardAll = "ca-app-pub-9507635869843997/3940067519"

    init {
        instantiateList()
        loadInitialRewards(context)
    }

    private fun instantiateList() {
        adUnits = ArrayList()
        adUnits!!.add(0, ArrayList(listOf(reward5, Stack<RewardedAd>())))
        adUnits!!.add(1, ArrayList(listOf(reward4, Stack<RewardedAd>())))
        adUnits!!.add(2, ArrayList(listOf(rewardHigh, Stack<RewardedAd>())))
        adUnits!!.add(3, ArrayList(listOf(rewardMed, Stack<RewardedAd>())))
        adUnits!!.add(4, ArrayList(listOf(rewardAll, Stack<RewardedAd>())))
    }

    fun loadInitialRewards(context: Context?) {
        RewardedAdLoad(context, totalLevels)
    }

    fun getMediumAd(activity: Context?): RewardedAd? {
        return getRewardedAd(activity, 1)
    }

    fun getHighFloorAd(activity: Context?): RewardedAd? {
        return getRewardedAd(activity, 2)
    }

    fun getDefaultAd(activity: Context?): RewardedAd? {
        return getRewardedAd(activity, 0)
    }

    fun getRewardedAd(activity: Context?, maxLevel: Int): RewardedAd? {
        for (i in totalLevels downTo 0) {
            if (maxLevel > i) {
                break
            }
            val list = adUnits!![i]
            val adunitid = list[0] as String
            val stack = list[1] as Stack<RewardedAd>
            RewardedAdLoadSpecific(activity, adunitid, stack)
            if (stack != null && !stack.isEmpty()) {
                return stack.pop()
            }
        }
        return null
    }

    fun RewardedAdLoad(activity: Context?, level: Int) {
        if (level < 0) {
            return
        }
        if (adUnits!!.size < level) {
            return
        }
        val list = adUnits!![level]
        val adunitid = list[0] as String
        val stack = list[1] as Stack<RewardedAd>
        val adRequest = AdRequest.Builder().build()
        RewardedAd.load(activity, adunitid, adRequest, object : RewardedAdLoadCallback() {
            override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                super.onAdFailedToLoad(loadAdError)
                RewardedAdLoad(activity, level - 1)
            }

            override fun onAdLoaded(rewardedAd: RewardedAd) {
                super.onAdLoaded(rewardedAd)
                stack.push(rewardedAd)
            }
        })
    }

    fun RewardedAdLoadSpecific(activity: Context?, adUnitId: String?, stack: Stack<RewardedAd>?) {
        val adRequest = AdRequest.Builder().build()
        RewardedAd.load(activity, adUnitId, adRequest, object : RewardedAdLoadCallback() {
            override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                super.onAdFailedToLoad(loadAdError)
            }

            override fun onAdLoaded(rewardedAd: RewardedAd) {
                super.onAdLoaded(rewardedAd)
                stack!!.push(rewardedAd)
            }
        })
    }
}